### PR TITLE
[FLINK-31072] Introduce streaming-read-atomic to ensure UB and UA cannot be split

### PR DESCRIPTION
**Pull request by JingsongLi**
Currently, streaming source will be checkpoint in any time, this means UPDATE_BEFORE and UPDATE_AFTER can be split into two checkpoint.
Downstream can see intermediate state. This is weird in some cases.
So in this ticket, add streaming-read-atomic:
The option to enable return per iterator instead of per record in streaming read. This can ensure that there will be no checkpoint segmentation in iterator consumption.

### CODE REVIEW

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/FlinkConnectorOptions.java [hunk 1]
```diff
@@ -78,6 +78,14 @@ public class FlinkConnectorOptions {
                                     + "By default, if this option is not defined, the planner will derive the parallelism "
                                     + "for each statement individually by also considering the global configuration.");
 
+    public static final ConfigOption<Boolean> STREAMING_READ_ATOMIC =
+            ConfigOptions.key("streaming-read-atomic")
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription(
+                            "The option to enable return per iterator instead of per record in streaming read."
```
Comment (SteNicholas) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/FlinkConnectorOptions.java:86]: Could this add the usage scenarios of this config in the description?

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/RecordsFunction.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.flink.api.connector.source.SourceOutput;
+import org.apache.flink.connector.base.source.reader.RecordEmitter;
+import org.apache.flink.connector.base.source.reader.RecordsWithSplitIds;
+import org.apache.flink.connector.file.src.impl.FileRecords;
+import org.apache.flink.connector.file.src.reader.BulkFormat;
+import org.apache.flink.connector.file.src.util.RecordAndPosition;
+import org.apache.flink.table.data.RowData;
+
+/** Records construction and emitter in source. */
+public interface RecordsFunction<T> extends RecordEmitter<T, RowData, FileStoreSourceSplitState> {
```
Comment (SteNicholas) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/RecordsFunction.java:30]: Could the interface methods of `RecordsFunction` add the java doc to comment the interface definition?

Comment (JingsongLi) [reply flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/RecordsFunction.java:30]: Let's add this

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/RecordsFunction.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    static IterateRecordsFunction forIterate() {
+        return new IterateRecordsFunction();
+    }
+
+    static SingleRecordsFunction forSingle() {
+        return new SingleRecordsFunction();
+    }
+
+    /** Return one by one, possibly inserting checkpoint in the middle. */
+    class IterateRecordsFunction implements RecordsFunction<RecordAndPosition<RowData>> {
```
Comment (SteNicholas) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/RecordsFunction.java:52]: Is it need to introduce `RecordsFunctionFactory` to create the `RecordsFunction`?

Comment (JingsongLi) [reply flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/RecordsFunction.java:52]: No need

**Comment by SteNicholas** [comment]
@JingsongLi, could you please add a ITCase to verify whether to work as expected when setting `streaming-read-atomic` to true?

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/SingleIteratorRecords.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Override
+    public Set<String> finishedSplits() {
+        return finishedSplits;
+    }
+
+    @Override
+    public void recycle() {
+        if (recordsForSplit != null) {
+            recordsForSplit.releaseBatch();
```
Comment (SteNicholas) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/SingleIteratorRecords.java:87]: `recordsForSplit = null`

Comment (JingsongLi) [reply flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/SingleIteratorRecords.java:87]: No need to assign null here, it is illegal to recycle twice.

**Comment by JingsongLi** [comment]
> @JingsongLi, could you please add a ITCase to verify whether to work as expected when setting `streaming-read-atomic` to true?

Added

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Pull request merged**